system/metrics/sinks/prometheus/prometheusReporterService.cpp

```diff
@@ -0,0 +1,354 @@
+    response->setStatus(HTTP_STATUS_OK);
+
+    response->send();
+    return true;
+}
+
+const char * mapMockMetricTypeToPrometheusStr(MetricType type)
+{
+    switch (type)
+    {
+        case hpccMetrics::METRICS_COUNTER:
+            return "counter";
+        case hpccMetrics::METRICS_GAUGE:
+            return "gauge";
+        default:
+            WARNLOG("Encountered unknown metric - cannot map to Prometheus metric!");
+            return nullptr;
+    }
+}
+
+void PrometheusReporterService::toPrometheusMetrics(std::vector<std::shared_ptr<IMetric>> reportMetrics, StringBuffer & out)
+{
+    for (auto &pMetric: reportMetrics)
+    {
+        const std::string name = pMetric->queryName();
+        if (!pMetric->queryDescription().empty())
+			out.append("# HELP ").append(name.c_str()).append(" ").append(pMetric->queryDescription().c_str()).append("\n");
+
+		MetricType type = pMetric->queryMetricType();
+		const char * promtype = mapMockMetricTypeToPrometheusStr(type);
+		if (promtype && *promtype)
+			out.append("# TYPE ").append(name.c_str()).append(" ").append(promtype).append("\n");
+
+		out.append(name.c_str()).append(" ").append(std::to_string(pMetric->queryValue()).c_str()).append("\n");
+    }
+}
+
+bool PrometheusReporterService::onMetrics(CHttpRequest * request, CHttpResponse * response)
+{
+    StringBuffer payload;
+    if (!m_metricsReporter)
+        onMetricsUnavailable(request, response);
+    else
+    {
+        toPrometheusMetrics(m_metricsReporter->queryMetricsForReport(std::string(m_metricsSinkName)), payload);
+
+        response->setContent(payload.length(), payload.str());
+        response->setContentType(DEFAULT_PROMETHEUS_METRICS_SERVICE_RESP_TYPE);
+
+        response->send();
+    }
+
+    return true;
+}
+
+bool PrometheusReporterService::onPost(CHttpRequest * request, CHttpResponse * response)
+{
+    IEspContext& context = *request->queryContext();
+    DBGLOG("PrometheusReporterService::onPost");
+
+    return onPostNotFound(response, "");
+}
+
+bool PrometheusReporterService::onGet(CHttpRequest * request, CHttpResponse * response)
+{
+    DBGLOG("PrometheusReporterService::onGet");
+
+    response->setVersion(HTTP_VERSION);
+    response->addHeader("Expires", "0");
+    response->setStatus(HTTP_STATUS_OK);
+    response->setContentType(DEFAULT_PROMETHEUS_METRICS_SERVICE_RESP_TYPE);
+
+    IEspContext& context = *request->queryContext();
+    const char * serviceName = context.queryServiceName(nullptr); //why does this method require in param?
+
+    if (!stricmp(serviceName, m_metricsServiceName.str()))
+        return onMetrics(request, response);
+
+    return onGetNotFound(response, serviceName);
+}
+
+void PrometheusReporterService::handleRequest(ISocket* clientSocket)
+{
+    Owned<CHttpRequest>         request;
+    Owned<CHttpResponse>        response;
+
+    request.setown(new CHttpRequest(*clientSocket));
+    IEspContext* ctx = createEspContext(createHttpSecureContext(request.get()));
```

Comment:
References to IEspContext seem wrong.  This code is not running within esp, and that looks like a significant chunk of functionality being pulled in.

Reply:
Agreed, unfortunately the CHttpRequest/CHttpResponse utilize that object to keep some generic, and some ESP specific context info - perhaps a bit of an overkill. It feels like there should be base httpcontext, extended by the IEspContext, but as is the ESPContext is referenced in the http req/resp logic.
This particular service def doesn't need to create the "httpSecureContext" and will pass nullptr instead.

Reply:
@afishbeck what is the best solution for this problem.  You added system/httplib for the Hashicorp vault secrets.  Should that be used?  Should we compile that into jlib somehow so that the code is shared?